google-cloud-firestore/lib/google/cloud/firestore/query.rb

```diff
@@ -0,0 +1,613 @@
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.limit(5).offset(10)
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def offset num
+          new_query = @query.dup
+          new_query ||= StructuredQuery.new
+
+          new_query.offset = num
+
+          Query.start new_query, parent_path, client
+        end
+
+        ##
+        # Limits a query to return a fixed number of results. If the current
+        # query already has a limit set, this will overwrite it.
+        #
+        # @param [Integer] num The maximum number of results to return.
+        #
+        # @return [Query] New query with `limit` called on it.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.offset(10).limit(5)
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def limit num
+          new_query = @query.dup
+          new_query ||= StructuredQuery.new
+
+          new_query.limit = Google::Protobuf::Int32Value.new(value: num)
+
+          Query.start new_query, parent_path, client
+        end
+
+        ##
+        # Starts query results at a particular document value. The result set
+        # will include the document specified by `values`.
+        #
+        # If the current query already has specified `start_at` or
+        # `start_after`, this will overwrite it.
+        #
+        # When the query is sent to the server, the `values` will be used in
+        # `order`.
+        #
+        # @param [Object] values The field value to start the query at.
+        #
+        # @return [Query] New query with `start_at` called on it.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.start_at("NYC")
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def start_at *values
+          new_query = @query.dup
+          new_query ||= StructuredQuery.new
+
+          values = values.flatten.map { |value| Convert.raw_to_value value }
```

Comment:
You need to special handle __name___ (or rather FieldPath.documentId()) to make sure that you pass a resource name.

Reply:
The field value here is still treated as a string though, right? The field name in `order` is what needs to support `__name__`, isn't it?